pkg/components/storage.go

```diff
@@ -121,25 +124,164 @@ func startCSIPlugin(ctx context.Context, cfg *config.Config, kubeconfigPath stri
 		klog.Warningf("Failed to apply deployment %v: %v", deploy, err)
 		return err
 	}
-	if err := assets.ApplyDaemonSets(ctx, ds, renderTemplate, renderParamsFromConfig(cfg, lvmdRenderParams), kubeconfigPath); err != nil {
-		klog.Warningf("Failed to apply daemonsets %v: %v", ds, err)
+	if err := assets.ApplyValidatingWebhookConfiguration(ctx, vwc, kubeconfigPath); err != nil {
+		klog.Warningf("Failed to apply validating webhook configuration %v: %v", vwc, err)
 		return err
 	}
-	if err := assets.ApplySCCs(ctx, scc, nil, nil, kubeconfigPath); err != nil {
-		klog.Warningf("Failed to apply sccs %v: %v", scc, err)
+	if err := assets.ApplyGeneric(ctx, lvmClusters, nil, map[string]interface{}{}, nil, kubeconfigPath); err != nil {
+		klog.Warningf("Failed to apply lvmcluster: %v", err)
 		return err
 	}
+
 	return nil
 }
+
+// loadCSIPluginConfig sets up a file watcher on the configuration directory. It first creates a cancellable context
+// and determines the directory and file path of the configuration file. It then checks if the configuration directory
+// and file exist. If the configuration file exists, it copies the user's configuration, otherwise, it copies the default configuration.
+//
+// The function then creates a new file watcher and adds the configuration directory to the watcher's watch list.
+// It listens for file events in a separate goroutine. If the configuration file is modified, it copies the user's configuration again.
+// If the configuration file is removed, it copies the default configuration. If the file permissions are changed, it logs a warning.
+//
+// The function is designed to ensure that lvms always uses the most recent configuration through its hot-reload mechanism,
+// whether it's user-defined or the default, and reacts to changes in the configuration file in real-time.
+func loadCSIPluginConfig(ctx context.Context,
+	// usrCfg is the user's configuration file.
+	usrCfg string,
+	runtimeCfg string,
+) (*lvmd.Lvmd, error) {
+	usrCfgDir := filepath.Dir(usrCfg)
+	ctx, cancel := context.WithCancelCause(ctx)
+
+	// check if file exists, otherwise the watcher errors
+	fi, err := os.Stat(usrCfgDir)
+	if err != nil {
+		return nil, fmt.Errorf("config directory %s cannot be watched: %v", usrCfgDir, err)
+	} else if !fi.IsDir() {
+		return nil, fmt.Errorf("config directory %s is not a directory", usrCfgDir)
+	}
+
+	if _, err := os.Stat(usrCfg); err == nil {
+		copyUserCfg(usrCfg, runtimeCfg, cancel)
+	} else {
+		copyDefaultCfg(runtimeCfg, cancel)
+	}
+
+	// Create new watcher.
+	go func(ctx context.Context) {
+		watcher, err := fsnotify.NewWatcher()
+		defer func() {
+			if err := watcher.Close(); err != nil {
+				klog.Errorf("unable to close file watcher: %v", err)
+			}
+		}()
+		if err != nil {
+			klog.Errorf("unable to set up file watcher: %v", err)
+		}
+		if err = watcher.Add(usrCfgDir); err != nil {
+			klog.Errorf("unable to add file path to watcher: %v", err)
```

Comment:
Should we terminate the goroutine in case of errors? and maybe cancel the context?
```suggestion
			klog.Errorf("unable to set up file watcher: %v", err)
            return
		}
		if err = watcher.Add(usrCfgDir); err != nil {
			klog.Errorf("unable to add file path to watcher: %v", err)
            return
```

Reply:
good catch, will do